feat(allocator/vec): remove `ManuallyDrop` wrapper

Dunqing opened the pull request:
Partially revert https://github.com/oxc-project/oxc/pull/6623, It still keeps the check for non-`drop` element of `Vec`

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9742?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9742** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9742?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9742?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9679** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9679?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9656** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9656?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9655** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9655?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9647** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9730** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9730?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9729** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9729?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9728** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9742).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_allocator/src/vec.rs:
```diff
@@ -261,8 +257,7 @@ impl<T: Hash> Hash for Vec<'_, T> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let inner = &*self.0;
-        f.debug_tuple("Vec").field(inner).finish()
+        f.debug_tuple("Vec").field(&self.0).finish()
```
Maybe we can simplify this method to just `self.0.fmt(f)`?

I'm not sure. Maybe that'd produce different output.

Dunqing replied:
Yes, I apply your suggestion, it is okay to produce different output.

overlookmotel replied:
Ah. But is it OK? Don't we rely on output being `Vec([...])` in some places (snapshots maybe)?

If it changes the output, let's go back to what you had before. If we want to change it, we can do that in a separate PR.

Dunqing replied:
Only a debug test fails, I am ok to do it in a separate PR.

overlookmotel replied:
Sorry my suggestion was a waste of time!

overlookmotel commented on crates/oxc_allocator/src/vec.rs:
```diff
@@ -35,7 +34,7 @@ use crate::{Allocator, Box};
 /// a [`Vec`] will refuse to compile if called with a [`Drop`] type.
 #[derive(PartialEq, Eq)]
-pub struct Vec<'alloc, T>(pub(crate) ManuallyDrop<InnerVec<'alloc, T>>);
+pub struct Vec<'alloc, T>(pub(crate) InnerVec<'alloc, T>);
```
I don't think `pub(crate)` is required any more either.

```suggestion
pub struct Vec<'alloc, T>(InnerVec<'alloc, T>);
```

overlookmotel reviewed (commented):
Apart from the 2 comments below, this looks good.

We definitely need to keep the `const { Self::ASSERT_T_IS_NOT_DROP }` assertions. `Vec` itself is non-drop, but these assertions also make sure that `T` (what you *put in* the `Vec`) is also non-drop.

This prevents memory leaks because otherwise there'd be nothing stopping you creating, for example, a `Vec<'a, std::string::String>`.

```rs
{
    let mut vec = Vec::new_in(&allocator);
    vec.push("foobar".to_string());
    // `vec` gets dropped here, but `Vec` doesn't implement `Drop`,
    // so it doesn't drop the `String`.
    // The 6 bytes of heap memory owned by the `String` is not reclaimed = memory leak.
}
```

`const { Self::ASSERT_T_IS_NOT_DROP }` is a compile-time check which prevents this. The example above will refuse to compile due to this assertion.

overlookmotel commented:
### Merge activity

* **Mar 14, 3:51 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
feat(allocator/vec): remove `ManuallyDrop` wrapper (#9742)

Partially revert https://github.com/oxc-project/oxc/pull/6623, It still keeps the check for non-`drop` element of `Vec`
